Add bronze quality scale to Growatt Server integration

johanzander opened the pull request:
## Proposed change

This PR adds Bronze tier quality scale certification to the Growatt Server integration.

Changes:
- Added `quality_scale.yaml` with all Bronze rules marked as `done` or `exempt`
- Updated `manifest.json` to declare `quality_scale: bronze`

All Bronze requirements are met:
- **Code**: Runtime data, unique IDs, entity naming, coordinator pattern
- **Configuration**: Config flow with test coverage, unique config entries  
- **Testing**: Test before configure/setup with 95%+ coverage
- **Documentation**: Installation and removal instructions present

Two Bronze rules are exempt:
- `action-setup`: Integration does not register custom service actions
- `docs-actions`: Integration does not register custom service actions to document

## Bronze requirements

- [x] `action-setup` - Service actions are registered in async_setup (exempt - no custom actions)
- [x] `appropriate-polling` - If it's a polling integration, set an appropriate polling interval
- [x] `brands` - Has branding assets available for the integration
- [x] `common-modules` - Place common patterns in common modules
- [x] `config-flow-test-coverage` - Full test coverage for the config flow
- [x] `config-flow` - Integration needs to be able to be set up via the UI
- [x] `dependency-transparency` - Dependency transparency
- [x] `docs-actions` - The documentation describes the provided service actions that can be used (exempt - no custom actions)
- [x] `docs-high-level-description` - The documentation includes a high-level description of the integration brand, product, or service
- [x] `docs-installation-instructions` - The documentation provides step-by-step installation instructions for the integration, including, if needed, prerequisites
- [x] `docs-removal-instructions` - The documentation provides removal instructions
- [x] `entity-event-setup` - Entity events are subscribed in the correct lifecycle methods
- [x] `entity-unique-id` - Entities have a unique ID
- [x] `has-entity-name` - Entities use has_entity_name = True
- [x] `runtime-data` - Use ConfigEntry.runtime_data to store runtime data
- [x] `test-before-configure` - Test a connection in the config flow
- [x] `test-before-setup` - Check during integration initialization if we are able to set it up correctly
- [x] `unique-config-entry` - Don't allow the same device or service to be able to be set up twice

## Type of change

- [X] Code quality improvements to existing code or addition of tests

## Checklist

- [X] I understand the code I am submitting and can explain how it works.
- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

johanzander committed:
Add Bronze quality scale to Growatt Server integration

This commit adds Bronze tier quality scale certification to the Growatt Server
integration by:

- Adding quality_scale.yaml with all Bronze rules marked as done or exempt
- Updating manifest.json to declare quality_scale: bronze

All Bronze requirements are met:
- Code: Runtime data, unique IDs, entity naming, coordinator pattern
- Configuration: Config flow with test coverage, unique config entries
- Testing: Test before configure/setup with 95%+ coverage
- Documentation: Installation instructions present

Two Bronze rules are exempt:
- action-setup: No custom service actions
- docs-actions: No custom service actions to document

joostlek reviewed (changes requested):
Please add all rules to the file

johanzander committed:
added all scales to file

joostlek commented on homeassistant/components/growatt_server/quality_scale.yaml:
```diff
@@ -0,0 +1,68 @@
+  appropriate-polling: done
+  brands: done
+  common-modules: done
+  config-flow-test-coverage: done
```
`test_password_auth_incorrect_login` can reuse `mock_growatt_classic_api` and repatch on the fly (this saves an indent)

johanzander replied:
I have adressed feedback on the tests in a separate PR: https://github.com/home-assistant/core/pull/156413

joostlek commented on homeassistant/components/growatt_server/quality_scale.yaml:
```diff
@@ -0,0 +1,68 @@
+  appropriate-polling: done
+  brands: done
+  common-modules: done
+  config-flow-test-coverage: done
```
I would also expect a happy test for password auth, we now only seem ot have `test_password_auth_incorrect_login` but that one has an error in the middle

johanzander replied:
I do belive we have happy paths:

Token Auth Happy Paths:   
  - test_token_auth_single_plant 
  -  test_token_auth_multiple_plants 

  Password Auth Happy Paths:
  - test_password_auth_single_plant 
  -  test_password_auth_multiple_plants 

Did you miss this or did you mean something else?

joostlek commented on homeassistant/components/growatt_server/quality_scale.yaml:
```diff
@@ -0,0 +1,68 @@
+  log-when-unavailable: todo
+  parallel-updates: done
+  reauthentication-flow: todo
+  test-coverage: todo
```
Not sure what `test_number_boundary_values` tests. We already snapshot the boundaries, so they are set, and if we really want to, I think you can merge this with pre-existing tests

joostlek commented on homeassistant/components/growatt_server/quality_scale.yaml:
```diff
@@ -0,0 +1,68 @@
+  log-when-unavailable: todo
+  parallel-updates: done
+  reauthentication-flow: todo
+  test-coverage: todo
```
Idem for `test_turn_on_switch_api_error`

joostlek reviewed (changes requested):
Let's put the things we agree on fixing in the comments in the file and merge it, and improve from there

johanzander commented on homeassistant/components/growatt_server/quality_scale.yaml:
```diff
@@ -0,0 +1,68 @@
+  # Silver
+  action-exceptions:
+    status: exempt
+    comment: Integration does not register custom service actions.
```
aha, i confused actions and services.

johanzander commented on homeassistant/components/growatt_server/quality_scale.yaml:
```diff
@@ -0,0 +1,68 @@
+  brands: done
+  common-modules: done
+  config-flow-test-coverage: done
+  config-flow: done
```
you are looking at the classic API that returns a list - the V1 equivalent returns a dict...

johanzander commented on homeassistant/components/growatt_server/quality_scale.yaml:
```diff
@@ -0,0 +1,68 @@
+  appropriate-polling: done
+  brands: done
+  common-modules: done
+  config-flow-test-coverage: done
```
This test validates the existing behavior/implementaion of the Classic API config flow:

  In the Classic API flow:
  1. User submits credentials in async_step_password_auth
  2. After successful login, flow proceeds to async_step_plant
  3. If plant_list fails in async_step_plant, the flow aborts

 The abort happens because the user has already submitted the password form, and the implementation doesn't store credentials to enable retry with an error message shown on the previous form.

This test was added to ensure config flow test coverage for this error path, not to change existing behavior.  My primary focus has been to add / improve on the V1 API as this is officially supported authentication method. 

There are lots of issues with the classic API and I would rather spend time implementing support for inverters in V1 so users can migrate to this and we at some point in the future can deprecate the classic API.

johanzander reviewed (commented):
A lot to chew on :-).

I created a separate PR for the feedback on the tests, I think those made sense: https://github.com/home-assistant/core/pull/156413.

I assume you want me to sum up a TODO list in the quality scale file of the relevant feedback and then we merge this file, is this correct?

johanzander committed:
Update quality_scale.yaml with review feedback

- Add code improvement comments from PR review
- Change action-setup from exempt to done (number/switch entities have actions)
- Change docs-actions from exempt to done (actions are documented)
- Change action-exceptions from exempt to done (proper error handling in place)
- Change entity-category from done to todo (sensors need categories assigned)
- Change entity-device-class from todo to done (device classes implemented)

johanzander committed:
Merge branch 'dev' into growatt-quality-scale

johanzander committed:
Merge branch 'dev' into growatt-quality-scale

johanzander committed:
fixes review comments

johanzander committed:
Merge branch 'dev' into growatt-quality-scale

johanzander commented on homeassistant/components/growatt_server/quality_scale.yaml:
```diff
@@ -0,0 +1,71 @@
+# 4. Add serial_number field to DeviceInfo in sensor, number, and switch platforms using device_id/serial_id
+# 5. Replace custom precision field with suggested_display_precision to preserve full data granularity
+# 6. Merge GrowattRequiredKeysMixin into GrowattSensorEntityDescription using kw_only=True
+# 7. Add EntityCategory.DIAGNOSTIC to temperature and other diagnostic sensors
```
ok, got it. Hope I got the todo and comments to the right sections now then.

joostlek committed:
Fix

joostlek committed:
Merge branch 'dev' into growatt-quality-scale
